mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java

```diff
@@ -124,16 +127,47 @@ private void calculateChanges(File projectDirectory, Configuration configuration
-            boolean wasConfigured = mavenProjectConfigurator.configureTestRunner(mavenProject.getModel());
-            if (wasConfigured) {
-                configuration.dump(mavenProject.getBasedir());
-                if (isFailedStrategyUsed()) {
-                    SurefireReportStorage.copySurefireReports(mavenProject.getModel());
+        final ConfigLookup lookUp = new ConfigLookup(session.getExecutionRootDirectory());
+        final boolean moreThanOneConfigFile = lookUp.hasMoreThanOneConfigFile(SMART_TESTING_YML, SMART_TESTING_YAML);
+        if (moreThanOneConfigFile && System.getProperty(SMART_TESTING_CONFIG) == null) {
+            session.getAllProjects().forEach(mavenProject -> {
+                ConfigLookup configLookup = new ConfigLookup(mavenProject.getBasedir().toString());
+                final File dirWithConfig = configLookup.getFirstDirWithConfigOrProjectRootDir();
+                Configuration mavenProjectConfiguration = configLookup.isConfigFromProjectRootDir() ? configuration :
+                    ConfigurationLoader.load(dirWithConfig);
+                if (mavenProjectConfiguration.isDisable()) {
+                    logExtensionDisableReason(logger,
+                        SMART_TESTING_DISABLE + " is set for module " + mavenProject.getArtifactId());
+                    return;
+                }
+                if (mavenProjectConfiguration.areStrategiesDefined()) {
+                    final MavenProjectConfigurator mavenProjectConfigurator =
+                        new MavenProjectConfigurator(mavenProjectConfiguration);
+                    configureMavenProject(mavenProjectConfigurator, mavenProject, mavenProjectConfiguration);
+                } else {
+                    logger.warn(
+                        "Smart Testing Extension is installed but no strategies are provided for %s module. It won't influence the way how your tests are executed. "
+                            + "For details on how to configure it head over to http://bit.ly/st-config",
+                        mavenProject.getArtifactId());
                 }
+            });
+        } else {
+            final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
+            session.getAllProjects().forEach(mavenProject -> {
+                configureMavenProject(mavenProjectConfigurator, mavenProject, configuration);
+            });
+        }
+    }
+
+    private void configureMavenProject(MavenProjectConfigurator mavenProjectConfigurator, MavenProject mavenProject,
+        Configuration configuration) {
+        boolean wasConfigured = mavenProjectConfigurator.configureTestRunner(mavenProject.getModel());
+        if (wasConfigured) {
+            configuration.dump(mavenProject.getBasedir());
+            if (isFailedStrategyUsed()) {
+                SurefireReportStorage.copySurefireReports(mavenProject.getModel());
             }
-        });
+        }
```

Comment:
My proposal for refactoring:

```java
private void configureExtension(MavenSession session, Configuration configuration) {
    final Consumer<MavenProject> configureSmartTestingExtensionAction;
    if (hasModuleSpecificConfigurations(session)) {
        configureSmartTestingExtensionAction = applyModuleSpecificConfiguration(configuration);
    } else {
        final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
        configureSmartTestingExtensionAction = mavenProject -> configureMavenProject(mavenProjectConfigurator, mavenProject, configuration);
    }
    session.getAllProjects().forEach(configureSmartTestingExtensionAction);
}

private boolean hasModuleSpecificConfigurations(MavenSession session) {
    final ConfigLookup lookUp = new ConfigLookup(session.getExecutionRootDirectory());
    final boolean moreThanOneConfigFile = lookUp.hasMoreThanOneConfigFile(SMART_TESTING_YML, SMART_TESTING_YAML);
    return moreThanOneConfigFile && System.getProperty(SMART_TESTING_CONFIG) == null;
}

private Consumer<MavenProject> applyModuleSpecificConfiguration(Configuration configuration) {
    return mavenProject -> {
        final ConfigLookup configLookup = new ConfigLookup(mavenProject.getBasedir().toString());
        final File dirWithConfig = configLookup.getFirstDirWithConfigOrProjectRootDir();
        final Configuration mavenProjectConfiguration = configLookup.isConfigFromProjectRootDir() ? configuration :
            ConfigurationLoader.load(dirWithConfig);
        if (mavenProjectConfiguration.isDisable()) {
            logExtensionDisableReason(logger,
                SMART_TESTING_DISABLE + " is set for module " + mavenProject.getArtifactId());
            return;
        }
        if (mavenProjectConfiguration.areStrategiesDefined()) {
            final MavenProjectConfigurator mavenProjectConfigurator =
                new MavenProjectConfigurator(mavenProjectConfiguration);
            configureMavenProject(mavenProjectConfigurator, mavenProject, mavenProjectConfiguration);
        } else {
            logger.warn(
                "Smart Testing Extension is installed but no strategies are provided for %s module. It won't influence the way how your tests are executed. "
                    + "For details on how to configure it head over to http://bit.ly/st-config",
                mavenProject.getArtifactId());
        }
    };
}
```

Reply:
To be honest, it took me some time to understand the code. as for assigning the consumer function to the local variable - I would use it only when the method returns some value. In this case I would use this:
```java 
    private void configureExtension(MavenSession session, Configuration configuration) {
        final File executionRootDirectory = new File(session.getExecutionRootDirectory());
        List<MavenProject> allProjects = session.getAllProjects();
        if (hasModuleSpecificConfigurations(executionRootDirectory, this::isProjectRootDirectory)) {
            allProjects.forEach(mavenProject -> applyModuleSpecificConfiguration(configuration, mavenProject));
        } else {
            final MavenProjectConfigurator mavenProjectConfigurator = new MavenProjectConfigurator(configuration);
            allProjects.forEach(mavenProject -> configureMavenProject(mavenProjectConfigurator, mavenProject, configuration));
        }
    }

    private void applyModuleSpecificConfiguration(Configuration configuration, MavenProject mavenProject) {
        final ConfigLookup configLookup = new ConfigLookup(mavenProject.getBasedir(), this::isProjectRootDirectory);
        final File dirWithConfig = configLookup.getFirstDirWithConfigOrWithStopCondition();
        final Configuration mavenProjectConfiguration = configLookup.isConfigFromProjectRootDir() ? configuration :
            ConfigurationLoader.load(dirWithConfig);
        if (mavenProjectConfiguration.isDisable()) {
            logExtensionDisableReason(logger,
                SMART_TESTING_DISABLE + " is set for module " + mavenProject.getArtifactId());
            return;
        }
        if (mavenProjectConfiguration.areStrategiesDefined()) {
            final MavenProjectConfigurator mavenProjectConfigurator =
                new MavenProjectConfigurator(mavenProjectConfiguration);
            configureMavenProject(mavenProjectConfigurator, mavenProject, mavenProjectConfiguration);
        } else {
            logger.warn(
                "Smart Testing Extension is installed but no strategies are provided for %s module. It won't influence the way how your tests are executed. "
                    + "For details on how to configure it head over to http://bit.ly/st-config",
                mavenProject.getArtifactId());
        }
    }
```
Just my opinion ;-)